Events removed when URL returns 5xx

### The problem

When the URL returns HTTP Error 5xx, Events are removed from the calendar.

Is it possible to do nothing to the calendar if requests are not successful?

### Version of GAS-ICS-Sync

5.7

### Additional information & file uploads

var howFrequent = 360;
var onlyFutureEvents = false;
var addEventsToCalendar = true;
var modifyExistingEvents = true;
var removeEventsFromCalendar = true;
var addAlerts = "no"; 
var addOrganizerToTitle = false; 
var descriptionAsTitles = false; 
var addCalToTitle = false; 
var addAttendees = false; 
var defaultAllDayReminder = -1; 
var overrideVisibility = "public";
var addTasks = false;

var emailSummary = true;

I just ran across this same issue last week.  I have an ical that is returning "[ERROR] Incorrect ics/ical URL..." and then all of the events are deleted.  I looked at #235 , but I hesitated to implement that code you gave Jonas because it looks like there are some conflicts with newer code in the script (colorID, etc.).  I also didn't really see where it was so different than what was in the Master in terms of handling the error (but you know my coding prowess is weak-sauce).

--- a/Code.gs
+++ b/Code.gs
@@ -145,6 +145,9 @@ var addedEvents = [];
 var modifiedEvents = [];
 var removedEvents = [];
 
+// Syncing logic can set this to true to cause the Google Apps Script "Executions" dashboard to report failure
+var reportOverallFailure = false;
+
 function startSync(){
   if (PropertiesService.getUserProperties().getProperty('LastRun') > 0 && (new Date().getTime() - PropertiesService.getUserProperties().getProperty('LastRun')) < 360000) {
     Logger.log("Another iteration is currently running! Exiting...");
@@ -249,4 +252,10 @@ function startSync(){
   }
   Logger.log("Sync finished!");
   PropertiesService.getUserProperties().setProperty('LastRun', 0);
+
+  if (reportOverallFailure) {
+    // Cause the Google Apps Script "Executions" dashboard to show a failure
+    // (the message text does not seem to be logged anywhere)
+    throw new Error('The sync operation produced errors. See log for details.');
+  }
 }

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -44,7 +44,7 @@ function formatDate(date) {
 
   const time = date.slice(11,16)
   const timeZone = date.slice(19)
-  
+
   return formattedDate + " at " + time + " (UTC" + (timeZone == "Z" ? "": timeZone) + ")"
 }
 
@@ -132,39 +132,45 @@ function fetchSourceCalendars(sourceCalendarURLs){
   for (var source of sourceCalendarURLs){
     var url = source[0].replace("webcal://", "https://");
     var colorId = source[1];
-    
-    callWithBackoff(function() {
-      var urlResponse = UrlFetchApp.fetch(url, { 'validateHttpsCertificates' : false, 'muteHttpExceptions' : true });
-      if (urlResponse.getResponseCode() == 200){
-        var icsContent = urlResponse.getContentText()
-        const icsRegex = RegExp("(BEGIN:VCALENDAR.*?END:VCALENDAR)", "s")
-        var urlContent = icsRegex.exec(icsContent);
-        if (urlContent == null){
-          // Microsoft Outlook has a bug that sometimes results in incorrectly formatted ics files. This tries to fix that problem.
-          // Add END:VEVENT for every BEGIN:VEVENT that's missing it
-          const veventRegex = /BEGIN:VEVENT(?:(?!END:VEVENT).)*?(?=.BEGIN|.END:VCALENDAR|$)/sg;
-          icsContent = icsContent.replace(veventRegex, (match) => match + "\nEND:VEVENT");
 
-          // Add END:VCALENDAR if missing
-          if (!icsContent.endsWith("END:VCALENDAR")){
-              icsContent += "\nEND:VCALENDAR";
-          }          
-          urlContent = icsRegex.exec(icsContent)
+    try {
+      callWithBackoff(function() {
+        var urlResponse = UrlFetchApp.fetch(url, { 'validateHttpsCertificates' : false, 'muteHttpExceptions' : true });
+        if (urlResponse.getResponseCode() == 200){
+          var icsContent = urlResponse.getContentText()
+          const icsRegex = RegExp("(BEGIN:VCALENDAR.*?END:VCALENDAR)", "s")
+          var urlContent = icsRegex.exec(icsContent);
           if (urlContent == null){
-            Logger.log("[ERROR] Incorrect ics/ical URL: " + url)
-            return
+            // Microsoft Outlook has a bug that sometimes results in incorrectly formatted ics files. This tries to fix that problem.
+            // Add END:VEVENT for every BEGIN:VEVENT that's missing it
+            const veventRegex = /BEGIN:VEVENT(?:(?!END:VEVENT).)*?(?=.BEGIN|.END:VCALENDAR|$)/sg;
+            icsContent = icsContent.replace(veventRegex, (match) => match + "\nEND:VEVENT");
+
+            // Add END:VCALENDAR if missing
+            if (!icsContent.endsWith("END:VCALENDAR")){
+                icsContent += "\nEND:VCALENDAR";
+            }
+            urlContent = icsRegex.exec(icsContent)
+            if (urlContent == null){
+              Logger.log("[ERROR] Incorrect ics/ical URL: " + url)
+              reportOverallFailure = true;
+              return
+            }
+            Logger.log("[WARNING] Microsoft is incorrectly formatting ics/ical at: " + url)
           }
-          Logger.log("[WARNING] Microsoft is incorrectly formatting ics/ical at: " + url)
+          result.push([urlContent[0], colorId]);
+          return;
         }
-        result.push([urlContent[0], colorId]);
-        return; 
-      }
-      else{ //Throw here to make callWithBackoff run again
-        throw "Error: Encountered HTTP error " + urlResponse.getResponseCode() + " when accessing " + url; 
-      }
-    }, defaultMaxRetries);
+        else{ //Throw here to make callWithBackoff run again
+          throw "Error: Encountered HTTP error " + urlResponse.getResponseCode() + " when accessing " + url;
+        }
+      }, defaultMaxRetries);
+    }
+    catch (e) {
+      reportOverallFailure = true;
+    }
   }
-  
+
   return result;
 }
 
@@ -242,7 +248,7 @@ function parseResponses(responses){
     });
   }
 
-  //No need to process calcelled events as they will be added to gcal's trash anyway
+  //No need to process cancelled events as they will be added to gcal's trash anyway
   result = result.filter(function(event){
     try{
       return (event.getFirstPropertyValue('status').toString().toLowerCase() != "cancelled");
@@ -451,7 +457,7 @@ function createEvent(event, calendarTz){
     if (organizerMail)
       newEvent.organizer.email = organizerMail.toString();
 
-    if (addOrganizerToTitle && organizerName){  
+    if (addOrganizerToTitle && organizerName){
         newEvent.summary = organizerName + ": " + newEvent.summary;
     }
   }
@@ -1085,7 +1091,14 @@ function sendSummary() {
 var backoffRecoverableErrors = [
   "service invoked too many times in a short time",
   "rate limit exceeded",
-  "internal error"];
+  "internal error",
+  "http error 403", // forbidden
+  "http error 408", // request timeout
+  "http error 423", // locked
+  "http error 500", // internal server error
+  "http error 503", // service unavailable
+  "http error 504"  // gateway timeout
+];
 function callWithBackoff(func, maxRetries) {
   var tries = 0;
   var result;
@@ -1097,10 +1110,7 @@ function callWithBackoff(func, maxRetries) {
     }
     catch(err){
       err = err.message  || err;
-      if ( err.includes("HTTP error") ) {
-        Logger.log(err);
-        return null;
-      } else if ( err.includes("is not a function")  || !backoffRecoverableErrors.some(function(e){
+      if ( err.includes("is not a function")  || !backoffRecoverableErrors.some(function(e){
               return err.toLowerCase().includes(e);
             }) ) {
         throw err;

